metricbeat/schema/schema.go

```diff
@@ -71,19 +71,13 @@ func (o Object) HasKey(key string) bool {
 
 // ApplyTo adds the fields extracted from data, converted using the schema, to the
 // event map.
-func (s Schema) ApplyTo(event common.MapStr, data map[string]interface{}) (common.MapStr, *Errors) {
```

Comment:
> advantage of having *Errors returned is that it forces the person implementing the code, to deal with it and understand the logic of it

I agree that the caller needs to deal with the error, but I don't think we should force them to understand the error. Returning a non-nil `error` should indicate failure only (so let's separate errors from warnings). If the caller wishes to inspect the error then there is a way to look at the individual errors caused by absent required fields.

Having a meta object to indicate the missing optional fields sounds like a good option to me. Basically we would separate errors from warnings. And warnings could be a separate return argument and distinct type. Outside of tests, I would expect most callers to be ignoring this meta return value with the list of absent optional fields (maybe with the exception of logging?).

If you agree, let's put this in the queue for #3807 and work it in a separate PR.